### PR TITLE
fresh new rules from al-khaser project

### PR DESCRIPTION
**Pull request by mike-hunhoff**
This PR contains a trove of new rules inspired by the [`al-khaser`](https://github.com/LordNoteworthy/al-khaser) project. There is a lot to unpack here and I'm happy to make any necessary changes, improvements, etc. prior to a merge.

I built 32-bit and 64-bit variants of the `al-khaser` project from source and used these binaries when writing the rules. I have also added the binaries to `capa-testfiles`.

There might also need to be a little deconfliction with #96.

Addressees #8.

### CODE REVIEW

File: anti-analysis/anti-vm/vm-detection/reference-anti-vm-strings.yml [hunk 1]
```diff
@@ -10,20 +10,43 @@ rule:
...
     examples:
       - Practical Malware Analysis Lab 17-02.dll_
   features:
     - or:
-      - string: /VMWare/i
-      - string: /VMTools/i
-      - string: /vmGuestLib.dll/i
-      - string: /VBox/i
-      - string: /VirtualBox/i
-      - string: /drivers\\vboxdrv/i
```
Comment (mike-hunhoff) [inline anti-analysis/anti-vm/vm-detection/reference-anti-vm-strings.yml:22]: These were moved to their respective "`targeting`" rules.

File: host-interaction/hardware/storage/get-disk-information.yml [hunk 2]
```diff
@@ -13,8 +13,6 @@ rule:
     - or:
       - api: kernel32.GetDriveType
       - api: kernel32.GetLogicalDrives
-      - api: kernel32.GetDiskFreeSpace
-      - api: kernel32.GetDiskFreeSpaceEx
```
Comment (mike-hunhoff) [inline host-interaction/hardware/storage/get-disk-information.yml:17]: There were moved to new rule `get-disk-size`.

File: linking/runtime-linking/access-peb-ldr_data.yml [hunk 3]
```diff
@@ -14,7 +14,7 @@ rule:
       - and:
         # resolve the PEB
         - or:
-          - characteristic: peb access
+          - match: PEB access
```
Comment (mike-hunhoff) [inline linking/runtime-linking/access-peb-ldr_data.yml:17]: change to `PEB access` match to detect multiple methods.

File: linking/runtime-linking/access-peb-ldr_data.yml [hunk 4]
```diff
@@ -37,7 +37,7 @@ rule:
...
```
Comment (mike-hunhoff) [inline linking/runtime-linking/access-peb-ldr_data.yml:40]: change to `PEB access` match to detect multiple methods.

File: anti-analysis/anti-debugging/debugger-detection/check-for-kernel-debugger-via-shared-user-data-structure.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+rule:
+  meta:
+    name: check for kernel debugger via shared user data structure
```
Comment (mike-hunhoff) [inline anti-analysis/anti-debugging/debugger-detection/check-for-kernel-debugger-via-shared-user-data-structure.yml:3]: maybe needs a better name?

File: anti-analysis/anti-debugging/debugger-detection/check-for-peb-beingdebugged-flag.yml [hunk 6]
```diff
@@ -12,5 +12,5 @@ rule:
       - Practical Malware Analysis Lab 16-01.exe_:0x403530
   features:
     - and:
-      - characteristic: peb access
+      - match: PEB access
```
Comment (mike-hunhoff) [inline anti-analysis/anti-debugging/debugger-detection/check-for-peb-beingdebugged-flag.yml:15]: change to `PEB access` match to detect multiple methods.

File: anti-analysis/anti-debugging/debugger-detection/check-for-peb-ntglobalflag-flag.yml [hunk 7]
```diff
@@ -3,7 +3,7 @@ rule:
     name: check for PEB NtGlobalFlag flag
     namespace: anti-analysis/anti-debugging/debugger-detection
     author: moritz.raabe@fireeye.com
-    scope: basic block
+    scope: function
```
Comment (mike-hunhoff) [inline anti-analysis/anti-debugging/debugger-detection/check-for-peb-ntglobalflag-flag.yml:6]: switched to `function` scope due to misses in `al-khaser` binaries.

File: anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml [hunk 8]
```diff
@@ -0,0 +1,43 @@
+rule:
+  meta:
+    name: reference anti-analysis strings targeting analysis tools
```
Comment (mike-hunhoff) [inline anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml:3]: maybe needs better name?

File: anti-analysis/anti-debugging/debugger-detection/check-for-debugger-via-api.yml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    scope: function
+    mbc:
+      - Anti-Behavioral Analysis::Detect Debugger
+    references:
+      - https://github.com/LordNoteworthy/al-khaser/blob/master/al-khaser/AntiDebug/CheckRemoteDebuggerPresent.cpp
+    examples:
+      - al-khaser_x86.exe_:0x420000
+  features:
+    - or:
+      - api: kernel32.IsDebuggerPresent
```
Comment (williballenthin) [inline anti-analysis/anti-debugging/debugger-detection/check-for-debugger-via-api.yml:15]: i think some exception handlers and other compiler-inserted code may add calls to `IsDebuggerPresent`, especially in debug builds. so, i think there's a chance for FP here (well, its a true match, but probably not what people are expecting). is this ok? should we remove `IsDebuggerPresent`? need more research?

thoughts, everyone?

File: anti-analysis/anti-debugging/debugger-detection/check-for-kernel-debugger-via-shared-user-data-structure.yml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    mbc:
+      - Anti-Behavioral Analysis::Detect Debugger
+    references:
+      - https://github.com/LordNoteworthy/al-khaser/blob/master/al-khaser/AntiDebug/SharedUserData_KernelDebugger.cpp
+      - http://www.geoffchappell.com/studies/windows/km/ntoskrnl/structs/kuser_shared_data.htm
+    examples:
+      - al-khaser_x86.exe_:0x430E60
+  features:
+    - and:
+      - number: 0x7FFE02D4 = UserSharedData->KdDebuggerEnabled
```
Comment (williballenthin) [inline anti-analysis/anti-debugging/debugger-detection/check-for-kernel-debugger-via-shared-user-data-structure.yml:16]: this is wild

File: anti-analysis/anti-debugging/debugger-detection/check-for-peb-ntglobalflag-flag.yml [hunk 11]
```diff
@@ -13,10 +13,20 @@ rule:
...
-        - offset: 0x68 = PEB.NtGlobalFlag
-        # 64-bit
-        - offset: 0xBC = PEB.NtGlobalFlag
+      - basic block:
+        - and:
+          - match: PEB access
+          - or:
+            - or:
+              # 32-bit
+              - offset: 0x68 = PEB.NtGlobalFlag
```
Comment (williballenthin) [inline anti-analysis/anti-debugging/debugger-detection/check-for-peb-ntglobalflag-flag.yml:22]: lets use arch flavors for this:
```yaml
or:
  - offset/x32: 0x68 = PEB.NtGlobalFlag
  - offset/x64: 0xBC = PEB.NEtGlobalFlag
```

File: anti-analysis/anti-debugging/debugger-detection/check-for-trap-flag-exception.yml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+      - Anti-Behavioral Analysis::Detect Debugger
+    references:
+      - https://github.com/LordNoteworthy/al-khaser/blob/master/al-khaser/AntiDebug/TrapFlag.cpp
+    examples:
+      - al-khaser_x86.exe_:0x431680
+      - al-khaser_x64.exe_:0x140030CB0
+  features:
+    - and:
+      - or:
+        # read/write EFLAGS
```
Comment (williballenthin) [inline anti-analysis/anti-debugging/debugger-detection/check-for-trap-flag-exception.yml:17]: you can use a statement description here.

File: anti-analysis/anti-debugging/debugger-detection/check-for-unexpected-memory-writes.yml [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    scope: basic block
+    mbc:
+      - Anti-Behavioral Analysis::Detect Debugger
+    references:
+      - https://github.com/LordNoteworthy/al-khaser/blob/master/al-khaser/AntiDebug/WriteWatch.cpp
+    examples:
+      - al-khaser_x86.exe_:0x431EBC
+  features:
+    - and:
+      - api: kernel32.GetWriteWatch
```
Comment (williballenthin) [inline anti-analysis/anti-debugging/debugger-detection/check-for-unexpected-memory-writes.yml:15]: what!?!

File: anti-analysis/anti-vm/vm-detection/reference-anti-vm-strings-targeting-virtualbox.yml [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+      - string: /82801FB/i
+      - string: /82441FX/i
+      - string: /82371SB/i
+      - string: /OpenHCD/i
+      - string: /ACPIBus_BUS_0/i
+      - string: /PCI_BUS_0/i
+      - string: /PNP_BUS_0/i
+      - string: /Oracle Corporation/i
+      - string: /VBoxWdd/i
+      - string: /VBoxS/i # VirtualBox Shared Folders
```
Comment (williballenthin) [inline anti-analysis/anti-vm/vm-detection/reference-anti-vm-strings-targeting-virtualbox.yml:68]: can use description block for this

File: anti-analysis/anti-vm/vm-detection/reference-anti-vm-strings-targeting-wine.yml [hunk 15]
```diff
@@ -0,0 +1,16 @@
+rule:
+  meta:
+    name: reference anti-VM strings targeting Wine
```
Comment (williballenthin) [inline anti-analysis/anti-vm/vm-detection/reference-anti-vm-strings-targeting-wine.yml:3]: merge with #95 ?

File: anti-analysis/anti-vm/vm-detection/reference-anti-vm-strings.yml [hunk 16]
```diff
@@ -10,20 +10,43 @@ rule:
...
+      - string: /KVMKVMKVM/i # KVM
+      - string: /Microsoft Hv/i # Microsoft Hyper-V or Windows Virtual PC
+      # https://github.com/LordNoteworthy/al-khaser/blob/master/al-khaser/AntiVM/Generic.cpp#L8
+      - string: /avghookx.dll/i # AVG
+      - string: /avghooka.dll/i # AVG
+      - string: /snxhk.dll/i # Avast
+      - string: /sbiedll.dll/i # Sandboxie
+      - string: /dbghelp.dll/i # WindBG
+      - string: /api_log.dll/i # iDefense Lab
+      - string: /dir_watch.dll/i # iDefense Lab
```
Comment (williballenthin) [inline anti-analysis/anti-vm/vm-detection/reference-anti-vm-strings.yml:38]: merge with #96 ?

File: host-interaction/hardware/cpu/get-number-of-processors.yml [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+    references:
+      - https://github.com/LordNoteworthy/al-khaser/blob/master/al-khaser/AntiVM/Generic.cpp#L113
+    examples:
+      - al-khaser_x86.exe_:0x432CB0
+  features:
+    - and:
+      - match: PEB access
+      - or:
+        - number: 0x64 # 32-bit
+        - number: 0xB8 # 64-bit
```
Comment (williballenthin) [inline host-interaction/hardware/cpu/get-number-of-processors.yml:18]: use arch flavors for this

**Comment by williballenthin** [comment]
couple comments inline, but overall, can't wait to merge this encyclopedia!

File: host-interaction/process/get-process-heap-flags.yml [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+      - https://github.com/LordNoteworthy/al-khaser/blob/master/al-khaser/AntiDebug/ProcessHeap_Flags.cpp
+    examples:
+      - al-khaser_x86.exe_:0x425470
+  features:
+    - and:
+      - match: PEB access
+      - or:
+        # 32-bit
+        - and:
+          - number/x32: 0x18 # offset process heap
```
Comment (Ana06) [inline host-interaction/process/get-process-heap-flags.yml:19]: what about adding these comment as descriptions (using `=`)?

**Comment by Ana06** [comment]
@github got crazy! It sent my review twice... 😨

**Comment by williballenthin** [comment]
@re-fox would be nice to have any feedback from you, if you have the time. no pressure, though.

File: anti-analysis/anti-debugging/debugger-detection/check-for-debugger-via-api.yml (hunk 9 above)
Comment (mr-tz) [reply anti-analysis/anti-debugging/debugger-detection/check-for-debugger-via-api.yml:15]: I think more research on the FP / unexpected rate match makes sense. In the meantime, I'd vote to remove it.

File: anti-analysis/anti-debugging/debugger-detection/check-for-software-breakpoints.yml [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+    references:
+      - https://github.com/LordNoteworthy/al-khaser/blob/master/al-khaser/AntiDebug/SoftwareBreakpoints.cpp
+    examples:
+      - al-khaser_x86.exe_:0x431020
+  features:
+    - and:
+      - basic block:
+        - and:
+          - mnemonic: cmp
+          - number: 0xCC
```
Comment (mr-tz) [inline anti-analysis/anti-debugging/debugger-detection/check-for-software-breakpoints.yml:18]: could add the less-used opcode `CD 03` 

`CD       ib  INT imm8                Interrupt vector number (imm8)`

File: anti-analysis/anti-debugging/debugger-detection/check-for-unexpected-memory-writes.yml (hunk 13 above)
Comment (mr-tz) [reply anti-analysis/anti-debugging/debugger-detection/check-for-unexpected-memory-writes.yml:15]: I've never seen this before, either 👍

File: anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml (hunk 8 above)
Comment (mr-tz) [reply anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml:3]: `reference analysis tools strings`?

File: anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+      - string: /sniff_hit.exe/i
+      - string: /windbg.exe/i
+      - string: /joeboxcontrol.exe/i
+      - string: /joeboxserver.exe/i
+      - string: /joeboxserver.exe/i
+      - string: /ResourceHacker.exe/i
+      - string: /x32dbg.exe/i
+      - string: /x64dbg.exe/i
+      - string: /Fiddler.exe/i
+      - string: /httpdebugger.exe/i
```
Comment (mr-tz) [inline anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml:43]: could add fakenet

File: lib/delay-execution.yml [hunk 21]
```diff
@@ -0,0 +1,25 @@
+rule:
+  meta:
+    name: delay execution
+    author: michael.hunhoff@fireeye.com
+    lib: true
```
Comment (mr-tz) [inline lib/delay-execution.yml:5]: this is a `lib` rule, but unused currently

File: lib/peb-access.yml [hunk 22]
```diff
@@ -0,0 +1,34 @@
+rule:
+  meta:
+    name: PEB access
```
Comment (mr-tz) [inline lib/peb-access.yml:3]: cool!

File: anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+      - string: /PETools.exe/i
+      - string: /LordPE.exe/i
+      - string: /SysInspector.exe/i
+      - string: /proc_analyzer.exe/i
+      - string: /sysAnalyzer.exe/i
+      - string: /sniff_hit.exe/i
+      - string: /windbg.exe/i
+      - string: /joeboxcontrol.exe/i
+      - string: /joeboxserver.exe/i
+      - string: /joeboxserver.exe/i
```
Comment (re-fox) [inline anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml:38]: `joeboxserver.exe` in here twice

File: anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml (hunk 20 above)
Comment (re-fox) [reply anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml:43]: another tool to add would be `WPE\ PRO.exe` and `netmon.exe`

File: host-interaction/hardware/storage/get-disk-size.yml [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+  features:
+    - or:
+      - api: kernel32.GetDiskFreeSpace
+      - api: kernel32.GetDiskFreeSpaceEx
+      - basic block:
+        - and:
+          - api: DeviceIoControl
+          - number: 0x7405C = IOCTL_DISK_GET_LENGTH_INFO
+      - and:
+        - string: /SELECT\s+\*\s+FROM\s+Win32_LogicalDisk/
```
Comment (re-fox) [inline host-interaction/hardware/storage/get-disk-size.yml:23]: Another variant that I've seen ITW is
```
SELECT * FROM Win32_DiskDrive WHERE (SerialNumber IS NOT NULL) AND (MediaType LIKE 'Fixed hard disk%')
```
which returns something like the following
```
Partitions : 2
DeviceID   : \\.\PHYSICALDRIVE0
Model      : xxxxxxxxxxxxxxxxx
Size       : 536268025360
Caption    : xxxxxxxxxxxxxxxxx
```

File: anti-analysis/anti-debugging/debugger-detection/check-for-debugger-via-api.yml (hunk 9 above)
Comment (mike-hunhoff) [reply anti-analysis/anti-debugging/debugger-detection/check-for-debugger-via-api.yml:15]: Good point - I've removed `IsDebuggerPresent` for now.

File: anti-analysis/anti-debugging/debugger-detection/check-for-peb-ntglobalflag-flag.yml (hunk 11 above)
Comment (mike-hunhoff) [reply anti-analysis/anti-debugging/debugger-detection/check-for-peb-ntglobalflag-flag.yml:22]: Good catch - addressed

File: anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml (hunk 23 above)
Comment (mike-hunhoff) [reply anti-analysis/reference-anti-analysis-strings-targeting-analysis-tools.yml:38]: Good catch, thank you!

**Pull request merged**